### PR TITLE
Rework GPU kernel policies to allow arbitrary problem size runs.

### PR DESCRIPTION
**Pull request by rhornung67**
This PR does the following:

1. modifies GPU policies for all nested loop kernels so that arbitrary problem sizes may be run
2. makes RAJA and non-RAJA variants mapped to blocks and threads consistent
3. adds checksum scaling to a bunch of kernels so that checksum differences should be < 1e-8 for all kernels to be considered correct.

### CODE REVIEW

File: src/basic/NESTED_INIT-Hip.cpp [hunk 1]
```diff
@@ -71,17 +96,18 @@ void NESTED_INIT::runHipVariant(VariantID vid)
...
+      dim3 nthreads_per_block(i_block_sz, j_block_sz, k_block_sz);
+      dim3 nblocks(static_cast<size_t>(RAJA_DIVIDE_CEILING_INT(ni, i_block_sz)),
+                   static_cast<size_t>(RAJA_DIVIDE_CEILING_INT(nj, j_block_sz)),
+                   static_cast<size_t>(RAJA_DIVIDE_CEILING_INT(nk, k_block_sz)));
 
-      auto kernel = lambda_hip_kernel<RAJA::hip_thread_x_direct, RAJA::hip_block_y_direct, RAJA::hip_block_z_direct, decltype(nested_init_lambda)>;
-      hipLaunchKernelGGL(kernel,
-        nblocks, nthreads_per_block, 0, 0,
-        0, ni, 0, nj, 0, nk, nested_init_lambda);
+      hipLaunchKernelGGL((nested_init_lam),
```
Comment (MrBurmark) [inline src/basic/NESTED_INIT-Hip.cpp:104]: This won't compile, `nested_init_lam` is just a function template, you have to give it `nested_init_lam<decltype(lambda)>` to get a function pointer.

Comment (rhornung67) [reply src/basic/NESTED_INIT-Hip.cpp:104]: Ah. Thanks. I hadn't gotten to the building and testing the Hip stuff yet. I was wondering about the seemingly overly complex code for Hip vs. CUDA.

Comment (MrBurmark) [reply src/basic/NESTED_INIT-Hip.cpp:104]: Another option is to use `(cuda|hip)LaunchKernel` for everything which would be consistent and have the least overhead.

Comment (rhornung67) [reply src/basic/NESTED_INIT-Hip.cpp:104]: I agree we should do that. But in a separate PR.

**Comment by artv3** [comment]
Regarding CUDA/HIP kernels, I think the RAJA variants impose __launch_bounds__ before the kernel launches which can help with optimizations. Would it be fair to add those in to the native variants for a more fair comparison?

**Comment by rhornung67** [comment]
> Regarding CUDA/HIP kernels, I think the RAJA variants impose **launch_bounds** before the kernel launches which can help with optimizations. Would it be fair to add those in to the native variants for a more fair comparison?

We should consider that, but let's work through the GPU kernel implementations first so that they are launched consistently across all GPU variants and we can run arbitrary problems sizes. I'm working through this now. Going more slowly than I would like due to meetings and other commitments.

File: src/polybench/POLYBENCH_2MM-Hip.cpp [hunk 2]
```diff
@@ -37,33 +57,60 @@ namespace polybench
...
+    POLYBENCH_2MM_BODY1;
+    for (Index_type k=0; k < nk; ++k) {
+      POLYBENCH_2MM_BODY2;
+    }
+    POLYBENCH_2MM_BODY3;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_2mm_1_lam(Index_type ni, Index_type nj, Index_type nk,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_2MM-Hip.cpp:77]: Looks like nk is unused here.

File: src/polybench/POLYBENCH_2MM-Hip.cpp [hunk 3]
```diff
@@ -37,33 +57,60 @@ namespace polybench
...
+    POLYBENCH_2MM_BODY4;
+    for (Index_type j=0; j < nj; ++j) {
+      POLYBENCH_2MM_BODY5;
+    }
+    POLYBENCH_2MM_BODY6;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_2mm_2_lam(Index_type ni,  Index_type nl, Index_type nj,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_2MM-Hip.cpp:105]: nj unused

File: src/polybench/POLYBENCH_2MM-Hip.cpp (hunk 2 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_2MM-Hip.cpp:77]: you are correct sir. nk will be captured by the lambda for that. same with cuda. I will push a fix shortly.

Comment (rhornung67) [reply src/polybench/POLYBENCH_2MM-Hip.cpp]: fixed.

File: src/polybench/POLYBENCH_2MM-Hip.cpp (hunk 3 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_2MM-Hip.cpp]: fixed.

File: src/polybench/POLYBENCH_3MM-Cuda.cpp [hunk 4]
```diff
@@ -42,44 +67,87 @@ namespace polybench
...
+    POLYBENCH_3MM_BODY1;
+    for (Index_type k=0; k < nk; ++k) {
+      POLYBENCH_3MM_BODY2;
+    }
+    POLYBENCH_3MM_BODY3;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_3mm_1_lam(Index_type ni, Index_type nj, Index_type nk,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_3MM-Cuda.cpp:85]: nk unused

File: src/polybench/POLYBENCH_3MM-Hip.cpp [hunk 5]
```diff
@@ -42,42 +67,84 @@ namespace polybench
...
+    POLYBENCH_3MM_BODY1;
+    for (Index_type k=0; k < nk; ++k) {
+      POLYBENCH_3MM_BODY2;
+    }
+    POLYBENCH_3MM_BODY3;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_3mm_1_lam(Index_type ni, Index_type nj, Index_type nk,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_3MM-Hip.cpp:85]: nk unused

File: src/polybench/POLYBENCH_GEMM-Cuda.cpp [hunk 6]
```diff
@@ -36,17 +51,31 @@ namespace polybench
...
+    POLYBENCH_GEMM_BODY2;
+    for (Index_type k = 0; k < nk; ++k ) {
+      POLYBENCH_GEMM_BODY3;
+    }
+    POLYBENCH_GEMM_BODY4;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_gemm_lam(Index_type ni, Index_type nj, Index_type nk,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_GEMM-Cuda.cpp:70]: nk unused

File: src/apps/LTIMES-Cuda.cpp [hunk 7]
```diff
@@ -34,14 +49,31 @@ namespace apps
...
+     for (Index_type d = 0; d < num_d; ++d ) {
+       LTIMES_BODY;
+     }
+   }
+}
 
-   for (Index_type d = 0; d < num_d; ++d ) {
-     LTIMES_BODY;
+template< typename Lambda >
+__global__ void ltimes_lam(Index_type num_d,
```
Comment (MrBurmark) [inline src/apps/LTIMES-Cuda.cpp:67]: num_d unused

File: src/apps/LTIMES-Hip.cpp [hunk 8]
```diff
@@ -34,14 +49,31 @@ namespace apps
...
+     for (Index_type d = 0; d < num_d; ++d ) {
+       LTIMES_BODY;
+     }
+   }
+}
 
-   for (Index_type d = 0; d < num_d; ++d ) {
-     LTIMES_BODY;
+template< typename Lambda >
+__global__ void ltimes_lam(Index_type num_d,
```
Comment (MrBurmark) [inline src/apps/LTIMES-Hip.cpp:67]: num_d unused

File: src/apps/LTIMES_NOVIEW-Cuda.cpp [hunk 9]
```diff
@@ -34,15 +50,32 @@ namespace apps
...
+       LTIMES_NOVIEW_BODY;
+     }
+   }
+}
 
-  for (Index_type d = 0; d < num_d; ++d ) {
-    LTIMES_NOVIEW_BODY;
-  }
+template< typename Lambda >
+__global__ void ltimes_noview_lam(Index_type num_d,
```
Comment (MrBurmark) [inline src/apps/LTIMES_NOVIEW-Cuda.cpp:68]: num_d unused

File: src/apps/LTIMES_NOVIEW-Hip.cpp [hunk 10]
```diff
@@ -34,14 +49,31 @@ namespace apps
...
+
+   if (m < num_m && g < num_g && z < num_z) {
+     for (Index_type d = 0; d < num_d; ++d ) {
+       LTIMES_NOVIEW_BODY;
+     }
+   }
+}
+
+template< typename Lambda >
+__global__ void ltimes_noview_lam(Index_type num_d,
```
Comment (MrBurmark) [inline src/apps/LTIMES_NOVIEW-Hip.cpp:67]: num_d unused

File: src/polybench/POLYBENCH_FLOYD_WARSHALL-Cuda.cpp [hunk 11]
```diff
@@ -34,10 +49,25 @@ __global__ void poly_floyd_warshall(Real_ptr pout, Real_ptr pin,
...
+  Index_type i = blockIdx.y * blockDim.y + threadIdx.y;
+  Index_type j = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if ( i < N && j < N ) { 
+    POLYBENCH_FLOYD_WARSHALL_BODY;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_floyd_warshall_lam(Index_type k,
```
Comment (MrBurmark) [inline src/polybench/POLYBENCH_FLOYD_WARSHALL-Cuda.cpp:61]: k unused

File: src/polybench/POLYBENCH_3MM-Cuda.cpp (hunk 4 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_3MM-Cuda.cpp]: fixed

File: src/polybench/POLYBENCH_3MM-Hip.cpp (hunk 5 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_3MM-Hip.cpp]: fixed

File: src/polybench/POLYBENCH_GEMM-Cuda.cpp (hunk 6 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_GEMM-Cuda.cpp]: fixed

File: src/apps/LTIMES-Cuda.cpp (hunk 7 above)
Comment (rhornung67) [reply src/apps/LTIMES-Cuda.cpp]: fixed

File: src/apps/LTIMES-Hip.cpp (hunk 8 above)
Comment (rhornung67) [reply src/apps/LTIMES-Hip.cpp]: fixed

File: src/apps/LTIMES_NOVIEW-Cuda.cpp (hunk 9 above)
Comment (rhornung67) [reply src/apps/LTIMES_NOVIEW-Cuda.cpp]: fixed

File: src/apps/LTIMES_NOVIEW-Hip.cpp (hunk 10 above)
Comment (rhornung67) [reply src/apps/LTIMES_NOVIEW-Hip.cpp]: fixed

File: src/polybench/POLYBENCH_FLOYD_WARSHALL-Cuda.cpp (hunk 11 above)
Comment (rhornung67) [reply src/polybench/POLYBENCH_FLOYD_WARSHALL-Cuda.cpp]: fixed

File: src/polybench/POLYBENCH_GEMM-Hip.cpp [hunk 12]
```diff
@@ -36,17 +51,31 @@ namespace polybench
...
+    POLYBENCH_GEMM_BODY2;
+    for (Index_type k = 0; k < nk; ++k ) {
+      POLYBENCH_GEMM_BODY3;
+    }
+    POLYBENCH_GEMM_BODY4;
+  }
+}
+
+template< typename Lambda >
+__global__ void poly_gemm_lam(Index_type ni, Index_type nj, Index_type nk,
```
[the remark this answers is not in the document]
Comment (rhornung67) [reply src/polybench/POLYBENCH_GEMM-Hip.cpp]: fixed

**Review by artv3 — APPROVED** [review]

**Comment by rhornung67** [comment]
@MrBurmark all of your comments have been addressed. If you find any more, please let me know.

**Pull request merged**